lib/src/Pos/Logic/Full.hs

```diff
@@ -0,0 +1,246 @@
+
+            postSscShares = postSscCommon
+                SharesMsg
+                (\(MCShares key _) -> key)
+                (\id tm -> MCShares id <$> tm ^. tmShares . at id)
+                (\(MCShares key shares) -> sscProcessShares key shares)
+
+            postSscVssCert = postSscCommon
+                VssCertificateMsg
+                (\(MCVssCertificate vc) -> getCertId vc)
+                (\id tm -> MCVssCertificate <$> lookupVss id (tm ^. tmCertificates))
+                (\(MCVssCertificate cert) -> sscProcessCertificate cert)
+
+            postSscCommon
+                :: ( Buildable err, Buildable contents )
+                => SscTag
+                -> (contents -> StakeholderId)
+                -> (StakeholderId -> TossModifier -> Maybe contents)
+                -> (contents -> ExceptT err m ())
+                -> KeyVal (Tagged contents StakeholderId) contents m
+            postSscCommon sscTag contentsToKey toContents processData = KeyVal
+                { toKey = pure . tagWith contentsProxy . contentsToKey
+                , handleInv = sscIsDataUseful sscTag . unTagged
+                , handleReq = \(Tagged addr) -> toContents addr . view ldModifier <$> sscRunLocalQuery ask
+                , handleData = \dat -> do
+                      let addr = contentsToKey dat
+                      -- [CSL-685] TODO: Add here malicious emulation for network
+                      -- addresses when TW will support getting peer address
+                      -- properly
+                      -- Stale comment?
+                      handleDataDo dat addr =<< shouldIgnorePkAddress addr
+                }
+              where
+                contentsProxy = (const Proxy :: (contents -> k) -> Proxy contents) contentsToKey
+                ignoreFmt =
+                    "Malicious emulation: data "%build%" for id "%build%" is ignored"
+                handleDataDo dat id shouldIgnore
+                    | shouldIgnore = False <$ logDebug (sformat ignoreFmt id dat)
+                    | otherwise = sscProcessMessage processData dat
+                sscProcessMessage sscProcessMessageDo dat =
+                    runExceptT (sscProcessMessageDo dat) >>= \case
+                        Left err -> False <$ logDebug (sformat ("Data is rejected, reason: "%build) err)
+                        Right () -> return True
+
+            -- TODO implement these proper
+            -- I do not know of an existing program which will do what we want:
+            -- check the block (header) in isolation; a coarse-but-fast test
+            -- to eliminate obviously bogus things. False positive possible,
+            -- but no false negatives, like a bloom filter.
+            checkBlockHeader = \_ -> pure True
+            checkBlock = \_ -> pure True
+
+            logic :: Logic m
+            logic = Logic {..}
+
+            runLogicLayer :: forall y . m y -> m y
+            runLogicLayer = identity
```

Comment:
What is this function supposed to do? I presume at some point it won't be `identity` (otherwise it's useless). What will it do?

Reply:
In the beginning I thought it would run the workers. Then I realized that won't be the case: logic and diffusion are to be passive, and not influence the control flow of the program. So really there are 3 ingredients in the model: logic, diffusion, and control. Diffusion uses logic, control uses diffusion and logic.

At present the full logic layer doesn't need to bracket any resources here. Compare at the full diffusion layer, which does a lot of bracketing in `runDiffusionLayer` (network transport, kademlia instance, time-warp node, outbound queue dequeue thread etc.). It's conceivable that the full logic layer could need something similar in the future, but I suppose it's only speculation.